lib/wab/impl/configuration.rb

```diff
@@ -1,105 +1,228 @@
-          )).merge(overrides)
+
+      attr_accessor :map
+      
+      def initialize(usage, options)
+        @map = {}
+        opts = OptionParser.new(usage)
+        config_file = nil
+        log_level = Logger::WARN
+        
+        opts.on('-c', '--config PATH', String, 'Configuration file.') { |c| config_file = c }
+        opts.on('-r', '--require LIBRARY', String, 'Require.')        { |r| require r }
+        opts.on('-v', '--verbose', 'Increase verbosity.')             { log_level += 1 }
+        opts.on('-h', '--help', 'Show this display.')                 { puts opts.help; Process.exit!(0) }
+
+        # Process command-line arguments and append them, in order, to an empty hash @map
+        add_options(opts, options)
+
+        opts.parse(ARGV)
+
+        # Move the @map sideways and replace with defaults.
+        command_line_map = @map
+        @map = {}
+        build_default_map(options)
+
+        # If a config file was specified load it and merge into @map.
+        @map = merge_map(@map, parse_config_file(config_file)) unless config_file.nil?
+
+        # Merge in the command line map.
+        @map = merge_map(@map, command_line_map) unless command_line_map.empty?
+      end
+
+      # Walks the options map and calls +opts.on+ for each option so that all
+      # are provided when +--help+ is called.
+      def add_options(opts, options, path='')
+        options.each_pair { |k,v|
+          next unless v.is_a?(Hash)
+          key_path = path.empty? ? k.to_s : "#{path}.#{k}"
+          if v.has_key?(:val)
+            default = v[:val]
+            if default.is_a?(Array)
+              opts.on(v[:short], "--#{key_path} PAIR", String, v[:doc]) { |val| arg_append(key_path, val, v[:parse]) }
+            else
+              opts.on(v[:short], "--#{key_path} VALUE", v[:type], "#{v[:doc]} Default: #{default}") { |val| set(key_path, val) }
+            end
+          else
+            add_options(opts, v, key_path)
+          end
+        }
+      end
+
+      # Appends an arg to an array in the configuration.
+      def arg_append(path, val, parse)
+        parts = val.split('=')
+        if 1 < parts.length
+          val = {}
+          parse.each_index { |i| val[parse[i].to_sym] = parts[i] }
         end
+        a = get(path)
+        if a.nil?
+          a = []
+          set(path, a)
+        end
+        a.push(val)
       end
 
+      # Builds a map from the default options passed in.
+      def build_default_map(options, path='')
+        options.each_pair { |k,v|
+          next unless v.is_a?(Hash)
+          key_path = path.empty? ? k.to_s : "#{path}.#{k}"
+          if v.has_key?(:val)
+            set(key_path, v[:val])
+          else
+            build_default_map(v, key_path)
+          end
+        }
+      end
+
+      # Recursive merge of other into prime.
+      def merge_map(prime, other)
+        prime.merge(other) { |key,val,oval|
```

Comment:
please rename to a more descriptive block parameter.. `oval` can be *mis-read* as `old_value` instead of `other_value`

Reply:
Renamed to nice long descriptive named.